web/v1/snowflake/sql-runner/sql/standard/00-setup/01-main/01-stored-procedures.sql

```diff
@@ -0,0 +1,426 @@
+          FROM ` + drv_src_ref;
+
+      var drv_delete_stmt = `
+          DELETE FROM ` + drv_trg_ref + `
+          WHERE ` + drv_join_condition + `
+            AND ` + drv_trg_ref + `.` + DRV_PARTITION_KEY + `>=(` + drv_lower_limit + `);`;
+
+      var drv_insert_stmt = `
+          INSERT INTO ` + drv_trg_ref + `
+            SELECT ` + drv_trg_columns + `
+            FROM ` + drv_src_ref + `;`;
+  }
+
+  if (STAGE_NXT === 'TRUE') {
+      var stg_trg_columns = list_cols(STG_TRG_SCHEMA,STG_TRG_TBL);
+      var stg_src_columns = list_cols(STG_SRC_SCHEMA,STG_SRC_TBL);
+      if (cols_check(stg_trg_columns, stg_src_columns) === false) {
+          throw Error("ERROR: column mismatch:" + STG_TRG_TBL + "," + STG_SRC_TBL);
+      }
+
+      var stg_trg_ref = `` + STG_TRG_SCHEMA + `.` + STG_TRG_TBL;
+      var stg_src_ref = `` + STG_SRC_SCHEMA + `.` + STG_SRC_TBL;
+      var stg_del_condition = `` + STG_JOIN_KEY + ` IN
+                              (SELECT ` + STG_JOIN_KEY + ` FROM ` + stg_src_ref + `)`;
+
+      var stg_delete_stmt = `
+          DELETE FROM ` + stg_trg_ref + `
+          WHERE ` + stg_del_condition + `;`;
+
+      var stg_insert_stmt = `
+          INSERT INTO ` + stg_trg_ref + `
+            SELECT ` + stg_trg_columns + `
+            FROM ` + stg_src_ref + `;`;
+  }
+
+  // BEGIN TRANSACTION
+  snowflake.createStatement({sqlText: `BEGIN;`}).execute();
+  try {
+
+      if (SKIP_DRV !== 'TRUE') {
+          snowflake.createStatement({sqlText: drv_delete_stmt}).execute();
+          snowflake.createStatement({sqlText: drv_insert_stmt}).execute();
+      }
+      if (STAGE_NXT === 'TRUE') {
+          snowflake.createStatement({sqlText: stg_delete_stmt}).execute();
+          snowflake.createStatement({sqlText: stg_insert_stmt}).execute();
+      }
+      snowflake.createStatement({sqlText: meta_stmt}).execute();
+
+      snowflake.createStatement({sqlText: `COMMIT;`}).execute();
+
+  } catch(ERROR) {
+
+      snowflake.createStatement({sqlText: `ROLLBACK;`}).execute();
+      throw ERROR;
+
+  }
+  return "ok. commit_with_metadata succeeded.";
+
+  // == Helpers ==
+
+  function list_cols(sch,tbl) {
+      var stmt = `
+          SELECT listagg(isc.column_name, ',') WITHIN GROUP (order by isc.ordinal_position)
+          FROM information_schema.columns AS isc
+          WHERE table_schema='` + sch + `'
+            AND table_name='` + tbl + `';`;
+
+      var res = snowflake.createStatement({sqlText: stmt}).execute();
+      res.next();
+      result = res.getColumnValue(1);
+
+      return result;
+  }
+
+  function cols_check(colsStr1, colsStr2) {
+      colsArr1 = colsStr1.split(',');
+      colsArr2 = colsStr2.split(',');
+
+      // since we know there are no duplicate cols in a table,just set comparison
+      return colsArr1.every(function(c) {return colsArr2.includes(c);}) &&
+             colsArr1.length === colsArr2.length;
+  }
+
+  $$
+;
+
+
+/*
+  CTAS_SIMILAR
```

Comment:
I like what you've done here - it's a clever solution. I wonder if we should consider trying to simplify things though - since we only need it once, do we need to create a procedure for this?

This is open for discussion, I'd like to think about it a bit, but one option is to just accept that base_staged has both `page_view_id` and `contexts_com_snowplowanalytics_snowplow_web_page_1`:

```
SELECT 
    contexts_com_snowplowanalytics_snowplow_web_page_1[0]:id AS page_view_id,
    *
FROM ...
```

Reply:
> since we only need it once, do we need to create a procedure for this?
Indeed, we do not need. But i actually thought that this may be the most useful procedure of all, maybe not to us or yet, but it makes it easy to slightly change a table and keep/drop/transform some columns. I can imagine a use case, but sure, no apparent need per se.

> one option is to just accept that base_staged has both page_view_id and contexts_com_snowplowanalytics_snowplow_web_page_1
This is indeed an option. But if we can do it, i wonder why not.